src/Compilers/CSharp/Portable/BoundTree/TupleBinaryOperatorInfo.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// A tree of binary operators for tuple comparisons.
+    ///
+    /// For `(a, (b, c)) == (d, (e, f))` we'll hold a Multiple with two elements.
+    /// The first element is a Single (describing the binary operator and conversions are involved in `a == d`).
+    /// The second element is a Multiple containing two Singles (one for the `b == e` comparison and the other for `c == f`).
+    /// </summary>
+    internal abstract class TupleBinaryOperatorInfo
+    {
+        internal abstract bool IsSingle();
+        internal readonly TypeSymbol LeftConvertedType;
+        internal readonly TypeSymbol RightConvertedType;
+#if DEBUG
+        internal abstract TreeDumperNode DumpCore();
+        internal string Dump() => TreeDumper.DumpCompact(DumpCore());
+#endif
+
+        private TupleBinaryOperatorInfo(TypeSymbol leftConvertedType, TypeSymbol rightConvertedType)
+        {
+            LeftConvertedType = leftConvertedType;
+            RightConvertedType = rightConvertedType;
+        }
+
+        /// <summary>
+        /// Holds the information for an element-wise comparison (like `a == b`)
+        /// </summary>
+        internal class Single : TupleBinaryOperatorInfo
+        {
+            internal readonly BinaryOperatorKind Kind;
+            internal readonly MethodSymbol MethodSymbolOpt; // User-defined comparison operator, if applicable
+
+            // To convert the result of comparison to bool
+            internal readonly Conversion BoolConversion;
+            internal readonly UnaryOperatorSignature BoolOperator; // Information for op_true or op_false
```

Comment:
>BoolOperator [](start = 53, length = 12)

What is expected in BoolConversion when this field is not null?

Reply:
As far as I can tell, `BoolConversion` will always be an identity conversion in such cases. That's the same as for regular equality. I kept the same logic for consistency and safety (just in case).

---
In reply to: [170489078](https://github.com/dotnet/roslyn/pull/23356#discussion_r170489078) [](ancestors = 170489078)

Reply:
> As far as I can tell, BoolConversion will always be an identity conversion in such cases. That's the same as for regular equality. I kept the same logic for consistency and safety (just in case).

It doesn't look like this is the case.

---
In reply to: [170741282](https://github.com/dotnet/roslyn/pull/23356#discussion_r170741282) [](ancestors = 170741282,170489078)